### DISCUSSION TITLE
Error during Build Next Auth v5 and Mongoose - Dynamic Code Evaluation

### DISCUSSION THREAD

**Discussion by ezefaz** [vercel/next.js#60212] [Help]
### Summary

Im having an error  related to the error "Dynamic Code Evaluation (e.g. 'eval', 'new Function', 'WebAssembly.compile') not allowed in Edge Runtime. 

Import trace for requested module:
./node_modules/mongoose/dist/browser.umd.js
./lib/models/user.model.ts
./auth.config.ts

" 

Im using Next-Auth V5 and Mongoose for the development. 

The issue happened when Im trying to get the User on the database. 

`  const user = await User.findOne({ email });`

// auth.config.ts

<img width="712" alt="image" src="https://github.com/vercel/next.js/assets/93097720/f0d7249a-e7d7-45fa-8820-fc239f06a759">

// user.model.ts

<img width="659" alt="image" src="https://github.com/vercel/next.js/assets/93097720/b9c06fbf-8469-4c05-9d22-400166a25bf5">

**Comment by Maliksidk19** [comment]
remove runtime=edge from route.ts, maybe it will work

**Reply by ezefaz** [reply]
solution was to add on the config the method unstable_allowDynamic and add the files that were dynamic.

export const config = {
  unstable_allowDynamic: [
    './auth.config.ts', 
    './lib/models/user.model.ts', 
    './node_modules/mongoose/dist/browser.umd.js', 
  ],
  
  Thanks for answering tho

**Reply by vikrant845** [reply]
![authError](https://github.com/vercel/next.js/assets/131388926/3a986e2e-72ac-4a4f-af0e-28ae13ac1ed1)
While in development mode i'm getting this error for the same code in auth.config.ts.

Any help would be appreciated

**Reply by ParsaJafarian** [reply]
Hi @ezefaz , in what file did you put the following code?
`
export const config = {
unstable_allowDynamic: [
'./auth.config.ts',
'./lib/models/user.model.ts',
'./node_modules/mongoose/dist/browser.umd.js',
],
`

**Reply by EEEcangorkemgunes** [reply]
Hi @vikrant845 ,
try to add ? after models. I mean;
export default mongoose.models?.User || mongoose.model<UserModelType>("User", userSchema);

**Reply by aanifowose111** [reply]
@ezefaz Could you please respond on where you added the config object you mentioned above. I have exact same issue you had.

**Comment by SameerSanap** [comment]
I am aslo in same situation. 
Did you find any solution to this ?

**Reply by Maliksidk19** [reply]
use this one 'just put a '?' question mark after the models so it check it the model exists then take that otherwise create new one'
`export const User = mongoose.models?.User || mongoose.model('User', userSchema)`

**Reply by SameerSanap** [reply]
hello sir, do you know about this error?
Failed to compile.

./node_modules\mongoose\dist\browser.umd.js
Dynamic Code Evaluation (e. g. 'eval', 'new Function', 'WebAssembly.compile') not allowed in Edge Runtime
Learn More: https://nextjs.org/docs/messages/edge-dynamic-code-evaluation

Import trace for requested module:
  ./node_modules\mongoose\dist\browser.umd.js
  ./utils\mongodb.js
  ./auth.js
  ./middleware.js


> Build failed because of webpack errors

**Reply by Maliksidk19** [reply]
show me you auth.js and middleware.js code, there might be something wrong in these files, as you can't use mongoose in edge functions

**Reply by v9shal** [reply]
hey , so i am also stuck around here,
@SameerSanap  did you find the solution for your solution

**Comment by fredle09** [comment]
I have the same problem with using `next-auth-v5`'s middleware, `mongoose`, and `next-intl`.

`./src\lib\auth.ts` file
```
// import libs
import NextAuth, {
  type Account,
  type User
} from "next-auth";
import bcrypt from "bcryptjs";

// import providers
import GitHub from "next-auth/providers/github";
import Facebook from "next-auth/providers/facebook";
import Credentials from "next-auth/providers/credentials";

// import models
import { connectToDb } from "./utils";
import { User as UserModel } from "./models";

// import config
import { authConfig } from "./auth.config";
import { Profile } from "next-auth";


interface ICredentials {
  username: string;
  password: string;
}

const login = async (credentials: ICredentials) => {
  try {
    // connectToDb();
    const user = await UserModel.findOne({ email: credentials.username });

    if (!user) throw new Error("Wrong credentials!");

    const isPasswordCorrect = await bcrypt.compare(
      credentials.password,
      user.password
    );

    if (!isPasswordCorrect)
      throw new Error("Wrong credentials!");

    return user;
  } catch (err) {
    console.log(err);
    throw new Error("Failed to login!");
  }
};

export const { handlers, signIn, signOut, auth } = NextAuth({
  providers: [
    GitHub({
      clientId: process.env.GITHUB_ID,
      clientSecret: process.env.GITHUB_SECRET,
    }),
    Facebook({
      clientId: process.env.FACEBOOK_ID,
      clientSecret: process.env.FACEBOOK_SECRET,
    }),
    Credentials({
      credentials: {
        username: { label: "Username", type: "text" },
        password: { label: "Password", type: "password" },
      },
      async authorize(credentials: ICredentials): Promise<User | null> {
        try {
          const user = await login(credentials);
          return user;
        } catch (err) {
          return null;
        }
      },
    }),
  ],
});
```

`./src\middleware.ts` file

```
import { NextResponse } from "next/server"
import type { NextRequest } from "next/server"
import createIntlMiddleware from "next-intl/middleware"

import { auth } from "@/lib/auth"

import { defaultLocale, localePrefix, locales } from "./navigation"

const privatePages: string[] = [
  // "/admin",
]

const authPages: string[] = [
  "/login",
  "/register"
]

const checkPathnameRegex = (pages: string[], pathName: string): boolean => {
  return RegExp(
    `^(/(${locales.join("|")}))?(${pages.flatMap((p) => (p === "/" ? ["", "/"] : p)).join("|")})/?$`,
    "i"
  ).test(pathName)
}

const intlMiddleware = createIntlMiddleware({
  locales,
  defaultLocale,
  localePrefix
})

const authMiddleware = auth((req) => {
  const isAuthPage = checkPathnameRegex(authPages, req.nextUrl.pathname);
  const session = req.auth;

  // Redirect to home page if authenticated and trying to access auth pages
  if (session && isAuthPage) {
    return NextResponse.redirect(new URL("/", req.nextUrl))
  }

  return intlMiddleware(req)
})

const middleware = (req: NextRequest) => {
  // const isPublicPage = checkPathnameRegex(publicPages, req.nextUrl.pathname)
  const isPrivatePage = checkPathnameRegex(privatePages, req.nextUrl.pathname)
  const isAuthPage = checkPathnameRegex(authPages, req.nextUrl.pathname)

  if (isAuthPage) {
    return (authMiddleware as any)(req)
  }

  if (!isPrivatePage) {
    return intlMiddleware(req)
  } else {
    return (authMiddleware as any)(req)
  }
}

export const config = {
  matcher: ["/((?!api|_next|.*\\..*).*)"],
}

export default middleware
```


the error when running `npm run build`
```
Failed to compile.

./node_modules\mongoose\dist\browser.umd.js
Dynamic Code Evaluation (e. g. 'eval', 'new Function', 'WebAssembly.compile') not allowed in Edge Runtime     
Learn More: https://nextjs.org/docs/messages/edge-dynamic-code-evaluation

Import trace for requested module:
  ./node_modules\mongoose\dist\browser.umd.js
  ./src\lib\models.ts
  ./src\lib\auth.ts
  ./src\middleware.ts


> Build failed because of webpack errors
```

I try to put this code in `next.config.js`
```
export const config = {
  unstable_allowDynamic: [
    "./node_modules\mongoose\dist\browser.umd.js",
    "./src\lib\models.ts",
    "./src\lib\auth.ts",
    "./src\middleware.ts",
  ],
}
```

But still not affected.

Did you find any solution to this?

**Comment by dharmveer97** [comment]
Does anyone have a solution for this one?

**Comment by shubham-kadam-brave** [comment]
Any updates?

**Reply by Maliksidk19** [reply]
just use prisma
